.github/workflows/npm-publish.yml

```diff
@@ -4,191 +4,238 @@ on:
+  NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
 
 jobs:
   publish_main:
     runs-on: ubuntu-latest
+    if: ${{ github.event.inputs.release_type == 'both' || github.event.inputs.release_type == 'main' }}
     steps:
-      - name: Setup environment
-        id: set-env
-        run: |
-          PUBLISH_ARGS=""
-          if [[ "${{ github.event.inputs.release_type }}" == "both" || "${{ github.event.inputs.release_type }}" == "main" ]]; then
-            echo "::set-output name=main::true"
-            PUBLISH_ARGS="$PUBLISH_ARGS --conventional-graduate"
-          elif [[ "${{ github.event.inputs.release_type }}" == "hotfix" ]]; then
-            echo "::set-output name=main::true"
-            echo "::set-output name=hotfix::true"
-          fi
-          if [[ "${{ github.event.inputs.force }}" == "true" ]]; then
-            PUBLISH_ARGS="$PUBLISH_ARGS --force-publish"
-          fi
-          if [[ -n $PUBLISH_ARGS ]]; then
-            echo "::set-output name=publishArgs::-- $PUBLISH_ARGS"
-          fi
-      - name: Setup Node
-        if: ${{ steps.set-env.outputs.main == 'true' }}
+      -
+        name: Setup Node
         uses: actions/setup-node@v1
         with:
           node-version: 16
-      - name: Checkout `main` branch
-        if: ${{ steps.set-env.outputs.main == 'true' }}
+      -
+        name: Checkout main branch
         uses: actions/checkout@v2
         with:
-          token: ${{ secrets.GH_TOKEN }}
           ref: main
-      - name: Configure git
-        if: ${{ steps.set-env.outputs.main == 'true' }}
+      -
+        name: Configure git
         run: |
           git config user.email "${{ github.actor }}@users.noreply.github.com"
           git config user.name ${{ github.actor }}
           git fetch --all
-      - name: Merge down from `main` -> `rc` -> `develop`
-        if: ${{ steps.set-env.outputs.main == 'true' }}
+      -
+        name: Merge down from main -> rc -> develop
         run: |
           git checkout -B release-candidate refs/remotes/origin/release-candidate
           git merge main
           git checkout -B develop refs/remotes/origin/develop
           git merge release-candidate
           git checkout main
-      - name: Merge `release-candidate` -> `main`
-        if: ${{ steps.set-env.outputs.main == 'true' && steps.set-env.outputs.hotfix != 'true' }}
+      -
+        name: Merge rc -> main
         run: git merge release-candidate
-      - name: Initialize and build code
-        if: ${{ steps.set-env.outputs.main == 'true' }}
+      -
+        name: Install Dagger
+        uses: dagger/dagger-for-github@v3
+        with:
+          install-only: true
+      -
+        name: Verify Docker image
+        run: |
+          dagger project init
+          dagger project update
+          dagger project update "github.com/3box/pipeline-tools/ci@v${{ env.PIPELINE_TOOLS_VER }}"
+          dagger do verify -p ${{ env.DAGGER_PLAN }}
+      -
+        name: Initialize and build code
```

Comment:
we don't need to build the code before the docker image?

Reply:
No, we build the code while making the Docker image, so not necessary. And this way, if verification fails for some reason, we won't have spent time building the code twice.

Reply:
annoying that the two steps can share the build, but oh well